### PR TITLE
Make change to TestDescriptor backwards compatible

### PR DESCRIPTION
**Pull request by ov7a**
Relates to
* #28019 

### Reviewing cheatsheet

Before merging the PR, comments starting with 
- ❌ ❓**must** be fixed
- 🤔 💅 **should** be fixed
- 💭 **may** be fixed
- 🎉 celebrate happy things

### CODE REVIEW

**Comment by ov7a** [comment]
@bot-gradle test RfR

**Comment by ov7a** [comment]
@bot-gradle test RfR

**Comment by ov7a** [comment]
@bot-gradle test RfN

File: platforms/software/testing-base/src/main/java/org/gradle/api/internal/tasks/testing/DecoratingTestDescriptor.java [hunk 1]
```diff
@@ -59,9 +59,12 @@ public String getClassName() {
         return descriptor.getClassName();
     }
 
-    @Override
     public String getMethodName() {
```
Comment (hegyibalint) [inline platforms/software/testing-base/src/main/java/org/gradle/api/internal/tasks/testing/DecoratingTestDescriptor.java:62]: 🤔 Class seems to be frozen as an exeption from `@NonNullApi`, consider removing the line and adding the annotations: https://github.com/gradle/gradle/blob/1a5e48701b602def428d9c79e3793adff7a4f798/testing/architecture-test/src/changes/archunit_store/internal-api-nullability.txt#L721

File: platforms/ide/tooling-api-builders/src/main/java/org/gradle/tooling/internal/provider/runner/TestOperationMapper.java [hunk 2]
```diff
@@ -84,12 +86,46 @@ public InternalOperationFinishedProgressEvent createFinishedEvent(DefaultTestDes
...
+    private static String getLegacyOperationDisplayName(String operationDisplayName, TestDescriptor testDescriptor) {
+        // if toString() is not overridden, use the display name for test operation
+        if (operationDisplayName.endsWith("@" + Integer.toHexString(testDescriptor.hashCode()))) {
+            return testDescriptor.getDisplayName();
+        } else {
+            return operationDisplayName;
+        }
+    }
+
+    // can be removed once the workaround above is removed
```
Comment (hegyibalint) [inline platforms/ide/tooling-api-builders/src/main/java/org/gradle/tooling/internal/provider/runner/TestOperationMapper.java:122]: ❗This comment is too loose on details: How can we know that the workaround is no more, so we could actually remove this? Also, is there an issue which would trigger this fix?

File: platforms/ide/tooling-api-builders/src/main/java/org/gradle/tooling/internal/provider/runner/TestOperationMapper.java [hunk 3]
```diff
@@ -84,12 +86,46 @@ public InternalOperationFinishedProgressEvent createFinishedEvent(DefaultTestDes
...
-        return new DefaultTestDescriptor(buildOperationId, test.getName(), test.toString(), test.getDisplayName(), InternalJvmTestDescriptor.KIND_ATOMIC, null, test.getClassName(), test.getName(), parentId, taskTracker.getTaskPath(buildOperationId));
+    // workaround for some plugins that override TestDescriptor
+    private static String getLegacyOperationDisplayName(String operationDisplayName, TestDescriptor testDescriptor) {
+        // if toString() is not overridden, use the display name for test operation
+        if (operationDisplayName.endsWith("@" + Integer.toHexString(testDescriptor.hashCode()))) {
+            return testDescriptor.getDisplayName();
+        } else {
+            return operationDisplayName;
+        }
+    }
```
Comment (hegyibalint) [inline platforms/ide/tooling-api-builders/src/main/java/org/gradle/tooling/internal/provider/runner/TestOperationMapper.java:120]: 💭 JIIUC, this is the hearth of the KGP fix.

**Review by hegyibalint — CHANGES_REQUESTED** [review]
Functionality LGTM, I would like to see an expanded comment about what is this workaround, if possible.

**Review by hegyibalint — APPROVED** [review]
I'm going to be OOO next week, and I don't want to complicate your life, so I will approve this, as the functionality is good.
Please still expand on the comment if you can, thanks!

File: platforms/ide/tooling-api-builders/src/main/java/org/gradle/tooling/internal/provider/runner/TestOperationMapper.java (hunk 2 above)
Comment (ghale) [reply platforms/ide/tooling-api-builders/src/main/java/org/gradle/tooling/internal/provider/runner/TestOperationMapper.java:122]: +1

**Review by ghale — APPROVED** [review]
I think this looks fine as long as we address the issues Balint points out.  Specifically that we document why this logic exists and what criteria would allow us to remove it in the future.

File: platforms/software/testing-base/src/main/java/org/gradle/api/tasks/testing/TestDescriptor.java [hunk 4]
```diff
@@ -52,16 +51,6 @@ public interface TestDescriptor {
...
 
-    /**
-     * Returns the method name for this test, if any.
-     *
-     * @return The method name. May return null.
-     * @since 8.8
-     */
-    @Incubating
-    @Nullable
-    String getMethodName();
```
[the remark this answers is not in the document]
Comment (ov7a) [reply platforms/software/testing-base/src/main/java/org/gradle/api/tasks/testing/TestDescriptor.java:63]: Moving the method from the interface to the abstract class allows to have backward-compatible interface overrides

**Pull request merged**